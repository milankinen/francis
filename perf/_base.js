const memwatch = require("@airbnb/node-memwatch")
const Benchmark = require("benchmark")
const bytes = require("pretty-bytes")
const { format } = require("util")
const { Francis, Bacon, Kefir } = require("./_libs")

const MEMORY = !!process.env.SHOW_MEMORY_USAGE
let heapStats = []
let gcTime = 0

const withFrancis = f => (...args) => f(Francis, ...args)
const withBacon = f => (...args) => f(Bacon, ...args)
const withKefir = f => (...args) => f(Kefir, ...args)

const resetHeapStats = () => {
  gcTime = 0
  heapStats = []
}

const getAvgMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => a + s, 0) / stats.length)) : "unknown"
const getMaxMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => Math.max(a, s), 0))) : "unknown"

if (MEMORY) {
  memwatch.on("stats", stats => {
    heapStats.push(stats.used_heap_size)
    gcTime += stats.gc_time / 1000000.0
  })
}

const testCases = (label, fn, combinations = []) => {
  const accepted = (testFn, args) => !testFn.accept || testFn.accept(...args)
  const cases = (name, testFn) =>
    combinations
      .map(
        args =>
          accepted(testFn, args) && {
            description:
              `${name} ${format(label + "|||", ...args).split("|||")[0]}` +
              (testFn.note ? ` (${testFn.note})` : ""),
            defer: true,
            fn: d => {
              const done = () => d.resolve()
              testFn(...args, done)
            },
          },
      )
      .filter(x => x)
  return [
    ...(process.env.NO_FRANCIS ? [] : cases("francis", withFrancis(fn))),
    ...(process.env.NO_BACON ? [] : cases("bacon  ", withBacon(fn))),
    ...(process.env.NO_KEFIR || !fn.kefir ? [] : cases("kefir  ", withKefir(fn.kefir))),
  ]
}

const runTests = (suiteName, testCases) => {
  return new Promise(resolve => {
    console.log(`Running suite ${suiteName}...`)
    console.log("")
    const maxLen = testCases.reduce((m, t) => Math.max(t.description.length, m), 0)
    resetHeapStats()
    testCases
      .reduce(
        (suite, test) => suite.add(test.description.padEnd(maxLen + 1), test),
        new Benchmark.Suite(),
      )
      .on("complete", function() {
        if (MEMORY) {
          const stats = [...heapStats]
          console.log("Memory statistics:")
          console.log(" > Avg mem:", getAvgMemoryUsage(stats))
          console.log(" > Max mem:", getMaxMemoryUsage(stats))
          console.log(" > GC time:", gcTime.toFixed(2), "ms")
        } else {
          this.forEach(test => console.log(test.toString()))
        }
        console.log("")
        resolve()
      })
      .run({ async: true })
  })
}

module.exports = {
  testCases,
  runTests,
}
