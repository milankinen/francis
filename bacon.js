export * from "./dist/francis.bacon"
