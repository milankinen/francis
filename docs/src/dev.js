import React from "react"
import { render } from "react-dom"
import { UI, setupInteractions } from "./ui"

const api = require("../api.json")
const intro = fetch(require("../INTRODUCTION.md")).then(x => x.text())

class DevUI extends React.Component {
  constructor(props) {
    super(props)
    this.state = {}
  }
  componentDidMount() {
    intro.then(i => this.setState({ intro: i }))
    setTimeout(setupInteractions, 0)
  }
  render() {
    const { intro } = this.state
    return intro ? <UI api={api} intro={intro} version="DEV" /> : null
  }
}

render(<DevUI />, document.getElementById("dev-app"))
