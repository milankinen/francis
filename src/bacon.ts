import { assert } from "./_assert"
import {
  Accum,
  Dispose,
  EndHandler,
  ErrorHandler,
  Handler,
  Predicate,
  Projection,
  ValueHandler,
  ValuesHandler,
} from "./_interfaces"
import {
  argsToObservables,
  argsToObservablesAndFunction,
  toFunction,
  toFunctionsPropAsIs,
} from "./_interrop"
import { EventStream } from "./EventStream"
import { Observable } from "./Observable"
import * as Buffer from "./operators/buffer"
import * as Changes from "./operators/changes"
import * as Combine from "./operators/combine"
import * as Concat from "./operators/concat"
import * as Do from "./operators/do"
import * as Errors from "./operators/errors"
import * as Filter from "./operators/filter"
import * as First from "./operators/first"
import * as FlatMap from "./operators/flatMap"
import * as Fold from "./operators/fold"
import * as Last from "./operators/last"
import * as Log from "./operators/log"
import * as Logic from "./operators/logic"
import * as Map from "./operators/map"
import * as MapEnd from "./operators/mapEnd"
import * as MapError from "./operators/mapError"
import * as Merge from "./operators/merge"
import * as Sample from "./operators/sample"
import * as Scan from "./operators/scan"
import * as Skip from "./operators/skip"
import * as SkipErrors from "./operators/skipErrors"
import * as SkipUntil from "./operators/skipUntil"
import * as SkipWhile from "./operators/skipWhile"
import * as StartWith from "./operators/startWith"
import * as Subscribe from "./operators/subscribe"
import * as Take from "./operators/take"
import * as TakeUntil from "./operators/takeUntil"
import * as TakeWhile from "./operators/takeWhile"
import * as Throttle from "./operators/throttle"
import * as ToEventStream from "./operators/toEventStream"
import * as ToProperty from "./operators/toProperty"
import * as Zip from "./operators/zip"
import { Property } from "./Property"
import { interval } from "./sources/interval"

declare module "./Observable" {
  interface Observable<A> {
    subscribe(handler: Handler<A>): Dispose
    onValue(f: ValueHandler<A>): Dispose
    onValues(f: ValuesHandler<A>): Dispose
    onError(f: ErrorHandler): Dispose
    onEnd(f: EndHandler): Dispose
    doAction(f: (val: A) => void): Observable<A>
    doError(f: (err: Error) => void): Observable<A>
    doEnd(f: () => void): Observable<A>
    doLog(label?: string): Observable<A>
    assign(obj: any, method: string, ...params: any[]): Dispose
    log(label?: string): Dispose
    map<B>(project: Projection<A, B>): Observable<B>
    mapError<A>(project: Projection<Error, A>): Observable<A>
    mapEnd(f: MapEnd.EndProjection<A>): Observable<A>
    filter(predicate: Predicate<A> | Property<any>): Observable<A>
    take(n: number): Observable<A>
    takeUntil(trigger: Observable<any>): Observable<A>
    takeWhile(f: Predicate<A> | Property<boolean>): Observable<A>
    first(): Observable<A>
    last(): Observable<A>
    skip(n: number): Observable<A>
    skipUntil(trigger: Observable<any>): Observable<A>
    skipWhile(trigger: Predicate<A> | Property<boolean>): Observable<A>
    skipErrors(): Observable<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): Observable<B>
    startWith(value: A): Observable<A>
    scan<B>(seed: B, f: Accum<B, A>): Property<B>
    fold<B>(seed: B, f: Accum<B, A>): Property<B>
    reduce<B>(seed: B, f: Accum<B, A>): Property<B>
    zip<B, C>(other: Observable<B>, f: (a: A, b: B) => C): EventStream<C>
    merge(other: Observable<A>): EventStream<A>
    concat(other: Observable<A>): EventStream<A>
    combine<B, C>(other: Observable<B>, f: (a: A, b: B) => C): Property<C>
    errors(): Observable<A>
    throttle(delay: number): Observable<A>
    bufferingThrottle(minimumInterval: number): Observable<A>
  }
}

declare module "./EventStream" {
  interface EventStream<A> {
    toProperty(initialValue?: A): Property<A>
    doAction(f: (val: A) => void): EventStream<A>
    doError(f: (err: Error) => void): EventStream<A>
    doEnd(f: () => void): EventStream<A>
    doLog(label?: string): EventStream<A>
    map<B>(project: Projection<A, B>): EventStream<B>
    mapError<A>(project: Projection<Error, A>): EventStream<A>
    mapEnd(f: A | MapEnd.EndProjection<A>): EventStream<A>
    filter(predicate: Predicate<A> | Property<any>): EventStream<A>
    take(n: number): EventStream<A>
    takeUntil(trigger: Observable<any>): EventStream<A>
    takeWhile(f: Predicate<A> | Property<boolean>): EventStream<A>
    first(): EventStream<A>
    last(): EventStream<A>
    skip(n: number): EventStream<A>
    skipUntil(trigger: Observable<any>): EventStream<A>
    skipWhile(trigger: Predicate<A> | Property<boolean>): EventStream<A>
    skipErrors(): EventStream<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): EventStream<B>
    startWith(value: A): EventStream<A>
    errors(): EventStream<A>
    throttle(delay: number): EventStream<A>
    bufferingThrottle(minimumInterval: number): EventStream<A>
    bufferWithTime(delay: number): EventStream<A[]>
    bufferWithCount(count: number): EventStream<A[]>
    bufferWithTimeOrCount(delay: number, count: number): EventStream<A[]>
  }
}

declare module "./Property" {
  interface Property<A> {
    toEventStream(): EventStream<A>
    changes(): EventStream<A>
    doAction(f: (val: A) => void): Property<A>
    doError(f: (err: Error) => void): Property<A>
    doEnd(f: () => void): Property<A>
    doLog(label?: string): Property<A>
    map<B>(project: Projection<A, B>): Property<B>
    mapError<A>(project: Projection<Error, A>): Property<A>
    mapEnd(f: MapEnd.EndProjection<A>): Property<A>
    filter(predicate: Predicate<A> | Property<any>): Property<A>
    take(n: number): Property<A>
    takeUntil(trigger: Observable<any>): Property<A>
    takeWhile(f: Predicate<A> | Property<boolean>): Property<A>
    first(): Property<A>
    last(): Property<A>
    skip(n: number): Property<A>
    skipUntil(trigger: Observable<any>): Property<A>
    skipWhile(trigger: Predicate<A> | Property<boolean>): Property<A>
    skipErrors(): Property<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): Property<B>
    sampledBy<B>(sampler: EventStream<B>): EventStream<A>
    sampledBy<B>(sampler: Property<B>): Property<A>
    sampledBy<B, C>(sampler: EventStream<B>, f: SampleFn<A, B, C>): EventStream<C>
    sampledBy<B, C>(sampler: Property<B>, f: SampleFn<A, B, C>): Property<C>
    sample(interval: number): EventStream<A>
    startWith(value: A): Property<A>
    and<B>(other: Property<B>): Property<Logic.AndResult<A, B>>
    or<B>(other: Property<B>): Property<Logic.OrResult<A, B>>
    not<B>(): Property<boolean>
    errors(): Property<A>
    throttle(delay: number): Property<A>
    bufferingThrottle(minimumInterval: number): Property<A>
  }
}

export type SampleFn<V, S, R> = (value: V, sample: S) => R

/* tslint:disable:no-string-literal */

// Observable operators (common for both EventStream and Property)

Observable.prototype["subscribe"] = function<A>(handler: Handler<A>, ...rest: any[]): Dispose {
  return Subscribe.subscribe(toFunction(handler, rest), this)
}

Observable.prototype["onValue"] = function<A>(f: ValueHandler<A>, ...rest: any[]): Dispose {
  return Subscribe.onValue(toFunction(f, rest), this)
}

Observable.prototype["onValues"] = function<A>(f: ValuesHandler<A>, ...rest: any[]): Dispose {
  return Subscribe.onValues(toFunction(f, rest), this)
}

Observable.prototype["onError"] = function<A>(f: ErrorHandler, ...rest: any[]): Dispose {
  return Subscribe.onError(toFunction(f, rest), this)
}

Observable.prototype["onEnd"] = function<A>(f: EndHandler, ...rest: any[]): Dispose {
  return Subscribe.onEnd(toFunction(f, rest), this)
}

Observable.prototype["assign"] = function<A>(obj: any, method: string, ...params: any[]): Dispose {
  return Subscribe.onValue(toFunction(obj, [method, ...params]), this)
}

Observable.prototype["doAction"] = function<A>(f: (val: A) => void, ...rest: any[]): Observable<A> {
  return Do.doAction(toFunction(f, rest), this)
}

Observable.prototype["doError"] = function<A>(
  f: (err: Error) => void,
  ...rest: any[]
): Observable<A> {
  return Do.doError(toFunction(f, rest), this)
}

Observable.prototype["doEnd"] = function<A>(f: () => void, ...rest: any[]): Observable<A> {
  return Do.doEnd(toFunction(f, rest), this)
}

Observable.prototype["doLog"] = function<A>(label?: string): Observable<A> {
  return Do.doLog(label, this)
}

Observable.prototype["log"] = function(label?: string): Dispose {
  return Log.log(label, this)
}

Observable.prototype["map"] = function<A, B>(
  project: Projection<A, B> | Property<B>,
  ...rest: any[]
): Observable<B> {
  return Map.map(toFunctionsPropAsIs(project, rest), this)
}

Observable.prototype["mapError"] = function<A>(
  project: Projection<Error, A>,
  ...rest: any[]
): Observable<A> {
  return MapError.mapError(toFunction(project, rest), this)
}

Observable.prototype["mapEnd"] = function<A>(f: MapEnd.EndProjection<A>): Observable<A> {
  return MapEnd.mapEnd(f, this)
}

Observable.prototype["filter"] = function<A>(
  predicate: Predicate<A> | Property<any>,
  ...rest: any[]
): Observable<A> {
  return Filter.filter(toFunctionsPropAsIs(predicate, rest), this)
}

Observable.prototype["take"] = function<A>(n: number): Observable<A> {
  return Take.take(n, this)
}

Observable.prototype["takeUntil"] = function<A>(trigger: Observable<any>): Observable<A> {
  return TakeUntil.takeUntil(trigger, this)
}

Observable.prototype["takeWhile"] = function<A>(
  f: Predicate<A> | Property<boolean>,
  ...rest: any[]
): Observable<A> {
  return TakeWhile.takeWhile(toFunctionsPropAsIs(f, rest), this)
}

Observable.prototype["skip"] = function<A>(n: number): Observable<A> {
  return Skip.skip(n, this)
}

Observable.prototype["skipUntil"] = function<A>(trigger: Observable<any>): Observable<A> {
  return SkipUntil.skipUntil(trigger, this)
}

Observable.prototype["skipWhile"] = function<A>(
  f: Predicate<A> | Property<boolean>,
  ...rest: any[]
): Observable<A> {
  return SkipWhile.skipWhile(toFunctionsPropAsIs(f, rest), this)
}

Observable.prototype["skipErrors"] = function<A>(): Observable<A> {
  return SkipErrors.skipErrors(this)
}

Observable.prototype["first"] = function<A>(): Observable<A> {
  return First.first(this)
}

Observable.prototype["last"] = function<A>(): Observable<A> {
  return Last.last(this)
}

Observable.prototype["flatMapLatest"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapLatest(toFunction(project, rest), this)
}

Observable.prototype["flatMapFirst"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapFirst(toFunction(project, rest), this)
}

Observable.prototype["flatMap"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMap(toFunction(project, rest), this)
}

Observable.prototype["flatMapConcat"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapConcat(toFunction(project, rest), this)
}

Observable.prototype["flatMapWithConcurrencyLimit"] = function<A, B>(
  limit: number,
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapWithConcurrencyLimit(limit, toFunction(project, rest), this)
}

Observable.prototype["startWith"] = function<A>(value: A): Observable<A> {
  return StartWith.startWith(value, this)
}

Observable.prototype["scan"] = function<A, B>(
  seed: B,
  f: Accum<B, A>,
  ...rest: any[]
): Property<B> {
  return Scan.scan(seed, toFunction(f, rest), this)
}

Observable.prototype["fold"] = Observable.prototype["reduce"] = function<A, B>(
  seed: B,
  f: Accum<B, A>,
  ...rest: any[]
): Property<B> {
  return Fold.fold(seed, toFunction(f, rest), this)
}

Observable.prototype["zip"] = function<A, B, C>(
  other: Observable<B>,
  f: (a: A, b: B) => C,
  ...rest: any[]
): EventStream<C> {
  return Zip.zipWith(toFunction(f, rest), [this, other])
}

Observable.prototype["merge"] = function<A>(other: Observable<A>): EventStream<A> {
  return Merge.mergeAll([this, other])
}

Observable.prototype["concat"] = function<A>(other: Observable<A>): EventStream<A> {
  return Concat.concatAll([this, other])
}

Observable.prototype["combine"] = function<A, B, C>(
  other: Observable<A>,
  f: (a: A, b: B) => C,
  ...rest: any[]
): Property<C> {
  return Combine.combineWith(toFunction(f, rest), [this, other] as any)
}

Observable.prototype["errors"] = function<A>(): Observable<A> {
  return Errors.errors(this)
}

Observable.prototype["throttle"] = function<A>(delay: number): Observable<A> {
  return Throttle.throttle(delay, this)
}

Observable.prototype["bufferingThrottle"] = function<A>(minimumInterval: number): Observable<A> {
  return Throttle.bufferingThrottle(minimumInterval, this)
}

// EventStream specific operators

EventStream.prototype["toProperty"] = function<A>(initialValue?: A): Property<A> {
  return arguments.length === 0
    ? ToProperty.toProperty(this)
    : StartWith._startWithP(initialValue, ToProperty.toProperty(this))
}

const esProto = EventStream.prototype as any
esProto["toEventStream"] = function<A>(): EventStream<A> {
  return this
}

EventStream.prototype["bufferWithTime"] = function<A>(delay: number): EventStream<A[]> {
  return Buffer.bufferWithTime(delay, this)
}

EventStream.prototype["bufferWithCount"] = function<A>(count: number): EventStream<A[]> {
  return Buffer.bufferWithCount(count, this)
}

EventStream.prototype["bufferWithTimeOrCount"] = function<A>(
  delay: number,
  count: number,
): EventStream<A[]> {
  return Buffer.bufferWithTimeOrCount(delay, count, this)
}

// Property specific operators

Property.prototype["sampledBy"] = function<A, B, C>(
  sampler: Observable<B>,
  f?: SampleFn<A, B, C>,
  ...rest: any[]
): any {
  const fn = toFunction(f === undefined ? (v: A, s: B) => v as any : f, rest)
  return Sample.sampleWith(sampler, fn, this)
}

Property.prototype["sample"] = function<A>(i: number): EventStream<A> {
  return Sample.sampleBy(interval(i, null as any), this)
}

Property.prototype["and"] = function<A, B>(other: Property<B>): Property<Logic.AndResult<A, B>> {
  return Logic.and(this, other)
}

Property.prototype["or"] = function<A, B>(other: Property<B>): Property<Logic.OrResult<A, B>> {
  return Logic.or(this, other)
}

Property.prototype["not"] = function<A>(): Property<boolean> {
  return Logic.not(this)
}

const pProto = Property.prototype as any
pProto["toProperty"] = function<A>(initialValue?: A): Property<A> {
  assert(arguments.length === 0, "No arguments supported")
  return this
}

Property.prototype["toEventStream"] = function<A>(): EventStream<A> {
  return ToEventStream.toEventStream(this)
}

Property.prototype["changes"] = function<A>(): EventStream<A> {
  return Changes.changes(this)
}

// static operators

export { End, Error, Next } from "./Event"
export { EventStream } from "./EventStream"
export { Observable } from "./Observable"
export { combineAsArray, combineTemplate } from "./operators/combine"
export { when } from "./operators/when"
export { zipAsArray } from "./operators/zip"
export { update } from "./operators/update"
export { Property } from "./Property"
export { fromArray } from "./sources/fromArray"
export { fromBinder } from "./sources/fromBinder"
export { fromEvent } from "./sources/fromEvent"
export { fromPoll } from "./sources/fromPoll"
export { interval } from "./sources/interval"
export { later } from "./sources/later"
export { never } from "./sources/never"
export { sequentially } from "./sources/sequentially"
export { constant, once } from "./sources/single"
export { repeat } from "./sources/repeat"
// classes and interfaces
export * from "./_interfaces"

export function zipWith<A, T>(f: (a: A) => T, streams: [Observable<A>]): EventStream<T>
export function zipWith<A, B, T>(
  f: (a: A, b: B) => T,
  streams: [Observable<A>, Observable<B>],
): EventStream<T>
export function zipWith<A, B, C, T>(
  f: (a: A, b: B, c: C) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>],
): EventStream<T>
export function zipWith<A, B, C, D, T>(
  f: (a: A, b: B, c: C, d: D) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>],
): EventStream<T>
export function zipWith<A, B, C, D, E, T>(
  f: (a: A, b: B, c: C, d: D, e: E) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>, Observable<E>],
): EventStream<T>
export function zipWith<A, B, C, D, E, F, T>(
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => T,
  streams: [
    Observable<A>,
    Observable<B>,
    Observable<C>,
    Observable<D>,
    Observable<E>,
    Observable<F>
  ],
): EventStream<T>
export function zipWith<A, T>(f: (...args: A[]) => T, streams: A[]): EventStream<T>
export function zipWith<A, T>(f: (streams: [Observable<A>], a: A) => T): EventStream<T>
export function zipWith<A, B, T>(
  streams: [Observable<A>, Observable<B>],
  f: (a: A, b: B) => T,
): EventStream<T>
export function zipWith<A, B, C, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>],
  f: (a: A, b: B, c: C) => T,
): EventStream<T>
export function zipWith<A, B, C, D, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>],
  f: (a: A, b: B, c: C, d: D) => T,
): EventStream<T>
export function zipWith<A, B, C, D, E, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>, Observable<E>],
  f: (a: A, b: B, c: C, d: D, e: E) => T,
): EventStream<T>
export function zipWith<A, B, C, D, E, F, T>(
  streams: [
    Observable<A>,
    Observable<B>,
    Observable<C>,
    Observable<D>,
    Observable<E>,
    Observable<F>
  ],
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => T,
): EventStream<T>
export function zipWith<A, T>(streams: A[], f: (...args: A[]) => T): EventStream<T>
export function zipWith(...args: any[]): EventStream<any> {
  const [observables, f] = argsToObservablesAndFunction(args)
  return Zip.zipWith(f, observables as any)
}

export function combineWith<A, R>(
  f: (a: A) => R,
  observables: [Combine.Combineable<A>],
): Property<R>
export function combineWith<A, B, R>(
  f: (a: A, b: B) => R,
  observables: [Combine.Combineable<A>, Combine.Combineable<B>],
): Property<R>
export function combineWith<A, B, C, R>(
  f: (a: A, b: B, c: C) => R,
  observables: [Combine.Combineable<A>, Combine.Combineable<B>, Combine.Combineable<C>],
): Property<R>
export function combineWith<A, B, C, D, R>(
  f: (a: A, b: B, c: C, d: D) => R,
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>
  ],
): Property<R>
export function combineWith<A, B, C, D, E, R>(
  f: (a: A, b: B, c: C, d: D, e: E) => R,
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>,
    Combine.Combineable<E>
  ],
): Property<R>
export function combineWith<A, B, C, D, E, F, R>(
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => R,
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>,
    Combine.Combineable<E>,
    Combine.Combineable<F>
  ],
): Property<R>
export function combineWith<A, R>(
  observables: [Combine.Combineable<A>],
  f: (a: A) => R,
): Property<R>
export function combineWith<A, B, R>(
  observables: [Combine.Combineable<A>, Combine.Combineable<B>],
  f: (a: A, b: B) => R,
): Property<R>
export function combineWith<A, B, C, R>(
  observables: [Combine.Combineable<A>, Combine.Combineable<B>, Combine.Combineable<C>],
  f: (a: A, b: B, c: C) => R,
): Property<R>
export function combineWith<A, B, C, D, R>(
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>
  ],
  f: (a: A, b: B, c: C, d: D) => R,
): Property<R>
export function combineWith<A, B, C, D, E, R>(
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>,
    Combine.Combineable<E>
  ],
  f: (a: A, b: B, c: C, d: D, e: E) => R,
): Property<R>
export function combineWith<A, B, C, D, E, F, R>(
  observables: [
    Combine.Combineable<A>,
    Combine.Combineable<B>,
    Combine.Combineable<C>,
    Combine.Combineable<D>,
    Combine.Combineable<E>,
    Combine.Combineable<F>
  ],
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => R,
): Property<R>
export function combineWith<T, R>(...args: any[]): Property<R> {
  const [observables, f] = argsToObservablesAndFunction(args)
  return Combine.combineWith(f, observables as any)
}

export function concatAll<T>(...observables: Array<Observable<T>>): EventStream<T>
export function concatAll<T>(observables: Array<Observable<T>>): EventStream<T>
export function concatAll<T>(
  observables: Array<Observable<T>> | Observable<T>,
  ...rest: Array<Observable<T>>
): EventStream<T> {
  return Concat.concatAll(argsToObservables([observables, ...rest]) as Array<EventStream<T>>)
}

export function mergeAll<T>(...observables: Array<Observable<T>>): EventStream<T>
export function mergeAll<T>(observables: Array<Observable<T>>): EventStream<T>
export function mergeAll<T>(
  observables: Array<Observable<T>> | Observable<T>,
  ...rest: Array<Observable<T>>
): EventStream<T> {
  return Merge.mergeAll(argsToObservables([observables, ...rest]))
}
