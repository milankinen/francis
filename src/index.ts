// interfaces and classes
export * from "./Event"
export * from "./_interfaces"
export { Observable } from "./Observable"
export { EventStream } from "./EventStream"
export { AnyObs, Property } from "./Property"

// factory functions
export { once } from "./sources/once"
export { constant } from "./sources/constant"
export { fromArray } from "./sources/fromArray"

// operators
export { map } from "./operators/map"
export { filter } from "./operators/filter"
export { take } from "./operators/take"
export { first } from "./operators/first"
export { zipAsArray } from "./operators/zip"
export { combineAsArray } from "./operators/combine"

// subscribers / side-effects
export { subscribe } from "./operators/subscribe"

export { log } from "./operators/log"
