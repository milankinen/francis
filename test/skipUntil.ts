import * as F from "../src/bacon"
import { recordActivationAndDispose, run, Sync } from "./_base"

describe("EventStream.skipUntil", () => {
  it("results in EventStream", () => {
    expect(F.once(1).skipUntil(F.never())).toBeInstanceOf(F.EventStream)
  })

  it("skips events until one appears in given trigger stream", () => {
    const recording = run(record =>
      F.sequentially(3, [1, 2, 3])
        .skipUntil(F.later(4, "<start>"))
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })

  it("works on self-derived trigger", () => {
    const recording = run(record => {
      const s = F.sequentially(3, [1, 2, 3])
      s.skipUntil(s.filter(x => x === 3)).subscribe(record)
    })
    expect(recording).toMatchSnapshot()
  })

  // "evil twist" :D
  it("works on self-derived stopper with an evil twist", () => {
    const recording = run(record => {
      const s = F.sequentially(3, [1, 2, 3])
      s.map(x => x)
        .take(3)
        .onValue(() => {})
      s.skipUntil(s.filter(x => x === 3)).subscribe(record)
    })
    expect(recording).toMatchSnapshot()
  })

  it("includes source errors, ignores trigger errors", () => {
    const recording = run(record => {
      F.sequentially(2, [1, new F.Error("oops" as any), 2, 3, 4])
        .skipUntil(F.sequentially(4, [new F.Error("tsers" as any), 1]))
        .subscribe(record)
    })
    expect(recording).toMatchSnapshot()
  })

  it("works with Property as trigger", () => {
    const recording = run(record =>
      F.sequentially(3, [1, 2, 3])
        .skipUntil(F.later(7, "<start>").toProperty())
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })

  it("considers Property init value as trigger", () => {
    const recording = run(record =>
      F.sequentially(3, [1, 2, 3])
        .skipUntil(F.later(7, "<start>").toProperty("tsers"))
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })

  it("ends properly when source ends", () => {
    const recording = run((record, wait) =>
      F.fromBinder(sink => {
        record("activate: source")
        wait(10, () => {
          sink(1)
          sink(2)
          sink(new F.End())
        })
        return () => {
          record("dispose: source")
        }
      })
        .skipUntil(recordActivationAndDispose(record, "trigger"))
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })

  it("end properly when trigger produces value", () => {
    const recording = run((record, wait) =>
      recordActivationAndDispose(record, "source")
        .merge(F.once("tsers"))
        .skipUntil(
          F.fromBinder(sink => {
            record("activate: trigger")
            wait(10, () => sink("stop!"))
            return () => record("dispose: trigger")
          }),
        )
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })
})

describe("Property.skipUntil", () => {
  it("results in Property", () => {
    expect(F.constant(1).skipUntil(F.never())).toBeInstanceOf(F.Property)
  })

  it("works synchronously", () => {
    const recording = run(record => {
      F.sequentially(3, [1, 2, 3])
        .toProperty(0)
        .skipUntil(F.later(7, "<start>").toProperty("tsers"))
        .subscribe(record),
        record(Sync)
    })
    expect(recording).toMatchSnapshot()
  })
})
