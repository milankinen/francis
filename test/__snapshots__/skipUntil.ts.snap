// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStream.skipUntil considers Property init value as trigger 1`] = `
Array [
  Next {
    "value": 1,
  },
  Next {
    "value": 2,
  },
  Next {
    "value": 3,
  },
  End {},
]
`;

exports[`EventStream.skipUntil end properly when trigger produces value 1`] = `
Array [
  "activate: trigger",
  "activate: source",
  "dispose: trigger",
]
`;

exports[`EventStream.skipUntil ends properly when source ends 1`] = `
Array [
  "activate: trigger",
  "activate: source",
  "dispose: source",
  "dispose: trigger",
  End {},
]
`;

exports[`EventStream.skipUntil includes source errors, ignores trigger errors 1`] = `
Array [
  Error {
    "error": "oops",
  },
  Next {
    "value": 3,
  },
  Next {
    "value": 4,
  },
  End {},
]
`;

exports[`EventStream.skipUntil skips events until one appears in given trigger stream 1`] = `
Array [
  Next {
    "value": 2,
  },
  Next {
    "value": 3,
  },
  End {},
]
`;

exports[`EventStream.skipUntil works on self-derived stopper with an evil twist 1`] = `
Array [
  Next {
    "value": 3,
  },
  End {},
]
`;

exports[`EventStream.skipUntil works on self-derived trigger 1`] = `
Array [
  Next {
    "value": 3,
  },
  End {},
]
`;

exports[`EventStream.skipUntil works with Property as trigger 1`] = `
Array [
  Next {
    "value": 3,
  },
  End {},
]
`;

exports[`Property.skipUntil works synchronously 1`] = `
Array [
  Sync {},
  Next {
    "value": 1,
  },
  Next {
    "value": 2,
  },
  Next {
    "value": 3,
  },
  End {},
]
`;
